### PR TITLE
Fill the correct throttle field by checking its frametype in offboard mode.

### PR DESCRIPTION
**Pull request by SalimTerryLi**
#12115

### CODE REVIEW

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 1]
```diff
@@ -1366,11 +1366,39 @@ MavlinkReceiver::handle_message_set_attitude_target(mavlink_message_t *msg)
...
-					// in that case we should assign att_sp.thrust_body[0]
 					if (!_offboard_control_mode.ignore_thrust) { // dont't overwrite thrust if it's invalid
-						att_sp.thrust_body[2] = -set_attitude_target.thrust;
+						// Fill correct field by checking frametype
+						// TODO: add as needed
+						switch (_mavlink->get_system_type()) {
+						case MAV_TYPE_GENERIC:
+							break;
+
+						case MAV_TYPE_FIXED_WING:
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_receiver.cpp]: Hm, so what happens for fixedwing?

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 2]
```diff
@@ -1366,11 +1366,39 @@ MavlinkReceiver::handle_message_set_attitude_target(mavlink_message_t *msg)
...
+							break;
+
+						case MAV_TYPE_QUADROTOR:
+							att_sp.thrust_body[2] = -set_attitude_target.thrust;
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_receiver.cpp]: Do you know why the quadcopter case has this `-`?

**Comment by julianoes** [comment]
@dagar could you have a look please?

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 1 above)
Comment (SalimTerryLi) [reply src/modules/mavlink/mavlink_receiver.cpp]: > 
> 
> Hm, so what happens for fixedwing?

I don' t know. Maybe we need someone focusing on fixed-wings?

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 2 above)
Comment (SalimTerryLi) [reply src/modules/mavlink/mavlink_receiver.cpp]: > 
> 
> Do you know why the quadcopter case has this `-`?

Sorry I don 't know ether. The controller of multi-copter is really a complex thing so I just treat it as black-box and just leave it there.

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 1 above)
Comment (SalimTerryLi) [reply src/modules/mavlink/mavlink_receiver.cpp]: > Hm, so what happens for fixedwing?

Attitude control from offboard mode makes no sense on a fixed-wing in my opinion. It should focus on velocity and position control instead. And it seems its controller only deals with `att_sp.thrust_body[0]` but ask for all of them.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 3]
```diff
@@ -1356,11 +1356,39 @@ MavlinkReceiver::handle_message_set_attitude_target(mavlink_message_t *msg)
...
-					// in that case we should assign att_sp.thrust_body[0]
 					if (!_offboard_control_mode.ignore_thrust) { // dont't overwrite thrust if it's invalid
-						att_sp.thrust_body[2] = -set_attitude_target.thrust;
+						// Fill correct field by checking frametype
+						// TODO: add as needed
+						switch (_mavlink->get_system_type()) {
+						case MAV_TYPE_GENERIC:
+							break;
+
+						case MAV_TYPE_FIXED_WING:
```
Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp]:
```suggestion
						case MAV_TYPE_FIXED_WING:
							att_sp.thrust_body[0] = set_attitude_target.thrust;
```

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 2 above)
Comment (dagar) [reply src/modules/mavlink/mavlink_receiver.cpp]: Positive z is down.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 4]
```diff
@@ -1356,11 +1356,39 @@ MavlinkReceiver::handle_message_set_attitude_target(mavlink_message_t *msg)
...
+							break;
+
+						case MAV_TYPE_QUADROTOR:
+							att_sp.thrust_body[2] = -set_attitude_target.thrust;
+							break;
+
+						case MAV_TYPE_HELICOPTER:
+							break;
```
Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp]: You can group all of the different rotary wing options together.

case MAV_TYPE_QUADROTOR:
case MAV_TYPE_HELICOPTER:
case MAV_TYPE_HEXAROTOR:
case MAV_TYPE_OCTOROTOR:
case MAV_TYPE_TRICOPTER:
    att_sp.thrust_body[2] = -set_attitude_target.thrust;

Comment (SalimTerryLi) [reply src/modules/mavlink/mavlink_receiver.cpp]: Helicopters as well? I think they have great difference, are n' t they?

Comment (dagar) [reply src/modules/mavlink/mavlink_receiver.cpp]: Yes, but in this context we're simply talking about Mavlink SET_ATTITUDE_TARGET (attitude + thrust). Right now we only need to distinguish between thrust down (helicopters, quadcopters, hex, etc) and thrust backwards (planes and cars).

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 5]
```diff
@@ -1356,11 +1356,31 @@ MavlinkReceiver::handle_message_set_attitude_target(mavlink_message_t *msg)
...
+
+						case MAV_TYPE_QUADROTOR:
+						case MAV_TYPE_HEXAROTOR:
+						case MAV_TYPE_OCTOROTOR:
+						case MAV_TYPE_TRICOPTER:
+							att_sp.thrust_body[2] = -set_attitude_target.thrust;
+							break;
+
+						case MAV_TYPE_HELICOPTER:
```
Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp]: Needs to be in the same group above (part of the FALLTHROUGH).

**Comment by Jaeyoung-Lim** [comment]
@SalimTerryLi I tried this with Fixedwing in SITL, but has no response on the throttle. Any ideas?

**Comment by SalimTerryLi** [comment]
> @SalimTerryLi I tried this with Fixedwing in SITL, but has no response on the throttle. Any ideas?
> 
> I would love to have offboard bodyrate control of fixed wing working again.

Maybe we can open a new issue about this? Rate control is probably disabled by [this](https://github.com/PX4/Firmware/blob/4831a4b0cfde499eb1a20257cc46b130e6554e3a/src/modules/mavlink/mavlink_receiver.cpp#L1422).

**Pull request merged**